main/src/ca/uwaterloo/flix/api/lsp/provider/completion/EnumTagCompleter.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2023 Lukas Rønn
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.api.lsp.provider.completion
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.api.lsp.Index
+import ca.uwaterloo.flix.api.lsp.provider.completion.Completion.EnumTagCompletion
+import ca.uwaterloo.flix.language.ast.{TypedAst, Symbol}
+
+object EnumTagCompleter extends Completer {
+
+  /**
+    * Returns a List of Completion for enum tags.
+    */
+  override def getCompletions(context: CompletionContext)(implicit flix: Flix, index: Index, root: TypedAst.Root, delta: DeltaContext): Iterable[EnumTagCompletion] = {
+    context.word.split('.').toList match {
+      /* We have the following cases:
+         0: We have nothing
+         1: We have path or enum
+         2: We have a path.enum, enum.tag, path.path
+            2.1: We have A.Color
+            2.2: We have Color.Green
+            2.3: We have A.B
+         3: We have an path.path.enum or path.enum.tag
+            3.1: We have A.B.Color
+            3.2: We have A.Color.Green
+         4: We have path.enum.tag i.e. A.B.Color.Green
+       */
+
+      // Case 0:
+      case Nil => Nil
+      // Case 1:
+      case x :: Nil =>
+        getEnumTagCompletions(Nil, x, None)
+      // Case 2:
+      case x :: y :: Nil =>
+        val ns1 = List(x)
+        // Case 2.1:
+        getEnumTagCompletions(ns1, y, None) ++
+          // Case 2.2:
+          getEnumTagCompletions(Nil, x, Some(y))
+          // Case 2.3:
+          // Do nothing, not valid for tags
+      // Case 3:
+      case x :: y :: z :: Nil =>
+        val ns1 = List(x)
+        val ns2 = List(x, y)
+        // Case 3.1:
+        getEnumTagCompletions(ns2, z, None) ++
+          // Case 3.2
+          getEnumTagCompletions(ns1, y, Some(z))
+      // Case 4:
+      case ns1 :: ns2 :: enum :: tag :: Nil =>
+        val ns = List(ns1, ns2)
+        getEnumTagCompletions(ns, enum, Some(tag))
+      case _ => Nil
+    }
+  }
```

Comment:
The above code strikes me as incredibly complicated. How is one supposed to reason about it?

Reply:
The whole aspect with nameSpace makes it incredibly complicated. I may have come up with a better and easier solution. Have to test it first.